File: 2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        def addTwoDigits(overflow_in:bool,n1:int,n2:int) ->(int,bool):
            tmp = n1 + n2
            if overflow_in:
                tmp += 1
            if tmp >= 10:
                return (tmp-10,True)
            else:
                return (tmp,False)
        ret_val, overflow = addTwoDigits(False,l1.val,l2.val)
        l1 = l1.next
        l2 = l2.next
        ret = ListNode(ret_val)
        cur = ret
        while l1 and l2:
            nex_val, overflow = addTwoDigits(overflow,l1.val,l2.val)
            cur.next = ListNode(nex_val)
            cur = cur.next
            l1 = l1.next
            l2 = l2.next
        if l1:
            tmp = l1
        elif l2:
            tmp = l2
        else:
            tmp = None
        while tmp:
            nex_val, overflow = addTwoDigits(overflow,tmp.val,0)
            cur.next = ListNode(nex_val)
            cur = cur.next
            tmp = tmp.next
        if overflow:
            cur.next = ListNode(1)
            overflow = False
        return ret

File: 2.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    var cur1, cur2 int
    res := &ListNode{
        Val: 0,
    }
    tmpPointer := res
    overflow := false
    for {
        if l1 == nil && l2 == nil {
            if overflow {
                tmpPointer.Val = 1
            }
            return res
        }
        if l1 != nil {
            cur1 = l1.Val
        } else {
            cur1 = 0
        }
        if l2 != nil {
            cur2  = l2.Val
        } else {
            cur2 = 0
        }

        var tmp int
        tmp, overflow = bitAdder(cur1, cur2, overflow)
        tmpPointer.Val = tmp

        if l1 != nil{
            l1 = l1.Next
        }
        if l2 != nil {
            l2 = l2.Next
        }
        if overflow || (l1 != nil) || (l2 != nil) {
            tmpPointer.Next = &ListNode{
                Val: 0,
            }
            tmpPointer = tmpPointer.Next
        } else {
            break
        }
    }
    return res
}

func bitAdder(b1, b2 int, bit bool) (sum int, overflow bool) {
    sum = b1+b2
    if bit {
        sum += 1
    }
    if sum > 9 {
        overflow = true
        sum -= 10
    }
    return
}
